tests fail on macOS

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
unit tests

##### ANSIBLE VERSION
devel branch (99e92f4bcad1498dba832f61f80b6b2014768391)


##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
- macOS Sierra 10.12.5
- python 2.7.13

##### SUMMARY
`make tests` fails with multiple errors.

##### STEPS TO REPRODUCE
```
# clone ansible
# setup virtualenv
pip install -r test/runner/requirements/units.txt
make tests
```

##### EXPECTED RESULTS
100% green tests

##### ACTUAL RESULTS
```
_____________________________________________________________________ TestModuleUtilsBasic.test_module_utils_basic_ansible_module_atomic_move _____________________________________________________________________

[...]

        if os.path.exists(b_dest):
            try:
                dest_stat = os.stat(b_dest)
    
                # copy mode and ownership
                os.chmod(b_src, dest_stat.st_mode & PERM_BITS)
                os.chown(b_src, dest_stat.st_uid, dest_stat.st_gid)
    
                # try to copy flags if possible
                if hasattr(os, 'chflags') and hasattr(dest_stat, 'st_flags'):
                    try:
>                       os.chflags(b_src, dest_stat.st_flags)
E                       TypeError: chflags() argument 2 must be an integer, not MagicMock

lib/ansible/module_utils/basic.py:2428: TypeError
_____________________________________________________________________________________ TestCollectedFacts.test_expected_facts ______________________________________________________________________________________

self = <units.module_utils.facts.test_ansible_collector.TestCollectedFacts testMethod=test_expected_facts>

    def test_expected_facts(self):
>       self._assert_expected_facts(self.facts)

test/units/module_utils/facts/test_ansible_collector.py:225: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/units/module_utils/facts/test_ansible_collector.py:252: in _assert_expected_facts
    self.assertIn(expected_fact, facts_keys)
E   AssertionError: 'distribution' not found in ['apparmor', 'architecture', 'date_time', 'dns', 'domain', 'effective_group_id', 'effective_user_id', 'env', 'fips', 'fqdn', 'gather_subset', 'hostname', 'kernel', 'local', 'lsb', 'machine', 'module_setup', 'nodename', 'pkg_mgr', 'python', 'python_version', 'real_group_id', 'real_user_id', 'selinux', 'selinux_python_present', 'system', 'user_dir', 'user_gecos', 'user_gid', 'user_id', 'user_shell', 'user_uid', 'userspace_architecture', 'userspace_bits', 'virtualization_role', 'virtualization_type']
---------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------
TypeError("'Mock' object is not iterable",)
TypeError("'Mock' object is not iterable",)
_________________________________________________________________________________ TestExceptionCollectedFacts.test_expected_facts _________________________________________________________________________________

self = <units.module_utils.facts.test_ansible_collector.TestExceptionCollectedFacts testMethod=test_expected_facts>

    def test_expected_facts(self):
>       self._assert_expected_facts(self.facts)

test/units/module_utils/facts/test_ansible_collector.py:225: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/units/module_utils/facts/test_ansible_collector.py:252: in _assert_expected_facts
    self.assertIn(expected_fact, facts_keys)
E   AssertionError: 'distribution' not found in ['apparmor', 'architecture', 'date_time', 'dns', 'domain', 'effective_group_id', 'effective_user_id', 'env', 'fips', 'fqdn', 'gather_subset', 'hostname', 'kernel', 'local', 'lsb', 'machine', 'module_setup', 'nodename', 'pkg_mgr', 'python', 'python_version', 'real_group_id', 'real_user_id', 'selinux', 'selinux_python_present', 'system', 'user_dir', 'user_gecos', 'user_gid', 'user_id', 'user_shell', 'user_uid', 'userspace_architecture', 'userspace_bits', 'virtualization_role', 'virtualization_type']
---------------------------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------------------------
Exception('A collector failed',)
TypeError("'Mock' object is not iterable",)
TypeError("'Mock' object is not iterable",)
[...]
```
... and many more. Most of them seem to be related to MagicMock.

--- a/test/lib/ansible_test/_data/requirements/constraints.txt
+++ b/test/lib/ansible_test/_data/requirements/constraints.txt
@@ -11,7 +11,7 @@ wheel < 0.30.0 ; python_version < '2.7' # wheel 0.30.0 and later require python 
 yamllint != 1.8.0, < 1.14.0 ; python_version < '2.7' # yamllint 1.8.0 and 1.14.0+ require python 2.7+
 pycrypto >= 2.6 # Need features found in 2.6 and greater
 ncclient >= 0.5.2 # Need features added in 0.5.2 and greater
-idna < 2.6 # requests requires idna < 2.6, but cryptography will cause the latest version to be installed instead
+idna < 2.6, >= 2.5 # linode requires idna < 2.9, >= 2.5, requests requires idna < 2.6, but cryptography will cause the latest version to be installed instead
 paramiko < 2.4.0 ; python_version < '2.7' # paramiko 2.4.0 drops support for python 2.6
 paramiko < 2.5.0 ; python_version >= '2.7' # paramiko 2.5.0 requires cryptography 2.5.0+
 pytest < 3.3.0 ; python_version < '2.7' # pytest 3.3.0 drops support for python 2.6

--- a/test/lib/ansible_test/_data/requirements/units.txt
+++ b/test/lib/ansible_test/_data/requirements/units.txt
@@ -10,6 +10,7 @@ pytest
 pytest-mock
 pytest-xdist
 python-memcached
+pytz
 pyvmomi
 pyyaml
 redis

--- a/test/units/module_utils/basic/test_atomic_move.py
+++ b/test/units/module_utils/basic/test_atomic_move.py
@@ -7,6 +7,7 @@
 from __future__ import absolute_import, division, print_function
 __metaclass__ = type
 
+import os
 import errno
 import json
 from itertools import product
@@ -27,7 +28,7 @@ def atomic_am(am, mocker):
 
 
 @pytest.fixture
-def atomic_mocks(mocker):
+def atomic_mocks(mocker, monkeypatch):
     environ = dict()
     mocks = {
         'chmod': mocker.patch('os.chmod'),
@@ -52,6 +53,9 @@ def atomic_mocks(mocker):
     mocks['getpwuid'].return_value = ('root', '', 0, 0, '', '', '')
     mocks['umask'].side_effect = [18, 0]
     mocks['rename'].return_value = None
+
+    # normalize OS specific features
+    monkeypatch.delattr(os, 'chflags')
 
     yield mocks
 

--- a/test/units/module_utils/basic/test_atomic_move.py
+++ b/test/units/module_utils/basic/test_atomic_move.py
@@ -55,7 +55,7 @@ def atomic_mocks(mocker, monkeypatch):
     mocks['rename'].return_value = None
 
     # normalize OS specific features
-    monkeypatch.delattr(os, 'chflags')
+    monkeypatch.delattr(os, 'chflags', raising=False)
 
     yield mocks
 

--- a/test/units/module_utils/basic/test_set_mode_if_different.py
+++ b/test/units/module_utils/basic/test_set_mode_if_different.py
@@ -101,19 +101,15 @@ def test_mode_unchanged_when_already_0660(am, mock_stats, mocker, mode, check_mo
 @pytest.mark.parametrize('check_mode, stdin',
                          product((True, False), ({},)),
                          indirect=['stdin'])
-def test_missing_lchmod_is_not_link(am, mock_stats, mocker, check_mode):
+def test_missing_lchmod_is_not_link(am, mock_stats, mocker, monkeypatch, check_mode):
     """Some platforms have lchmod (*BSD) others do not (Linux)"""
 
     am.check_mode = check_mode
     original_hasattr = hasattr
 
-    def _hasattr(obj, name):
-        if obj == os and name == 'lchmod':
-            return False
-        return original_hasattr(obj, name)
+    monkeypatch.delattr(os, 'lchmod')
 
     mocker.patch('os.lstat', side_effect=[mock_stats['before'], mock_stats['after']])
-    mocker.patch.object(builtins, 'hasattr', side_effect=_hasattr)
     mocker.patch('os.path.islink', return_value=False)
     mocker.patch('os.path.exists', return_value=True)
     m_chmod = mocker.patch('os.chmod', return_value=None)
@@ -128,19 +124,15 @@ def test_missing_lchmod_is_not_link(am, mock_stats, mocker, check_mode):
 @pytest.mark.parametrize('check_mode, stdin',
                          product((True, False), ({},)),
                          indirect=['stdin'])
-def test_missing_lchmod_is_link(am, mock_stats, mocker, check_mode):
+def test_missing_lchmod_is_link(am, mock_stats, mocker, monkeypatch, check_mode):
     """Some platforms have lchmod (*BSD) others do not (Linux)"""
 
     am.check_mode = check_mode
     original_hasattr = hasattr
 
-    def _hasattr(obj, name):
-        if obj == os and name == 'lchmod':
-            return False
-        return original_hasattr(obj, name)
+    monkeypatch.delattr(os, 'lchmod')
 
     mocker.patch('os.lstat', side_effect=[mock_stats['before'], mock_stats['after']])
-    mocker.patch.object(builtins, 'hasattr', side_effect=_hasattr)
     mocker.patch('os.path.islink', return_value=True)
     mocker.patch('os.path.exists', return_value=True)
     m_chmod = mocker.patch('os.chmod', return_value=None)

--- a/test/units/module_utils/basic/test_set_mode_if_different.py
+++ b/test/units/module_utils/basic/test_set_mode_if_different.py
@@ -107,7 +107,7 @@ def test_missing_lchmod_is_not_link(am, mock_stats, mocker, monkeypatch, check_m
     am.check_mode = check_mode
     original_hasattr = hasattr
 
-    monkeypatch.delattr(os, 'lchmod')
+    monkeypatch.delattr(os, 'lchmod', raising=False)
 
     mocker.patch('os.lstat', side_effect=[mock_stats['before'], mock_stats['after']])
     mocker.patch('os.path.islink', return_value=False)
@@ -130,7 +130,7 @@ def test_missing_lchmod_is_link(am, mock_stats, mocker, monkeypatch, check_mode)
     am.check_mode = check_mode
     original_hasattr = hasattr
 
-    monkeypatch.delattr(os, 'lchmod')
+    monkeypatch.delattr(os, 'lchmod', raising=False)
 
     mocker.patch('os.lstat', side_effect=[mock_stats['before'], mock_stats['after']])
     mocker.patch('os.path.islink', return_value=True)

--- a/test/units/module_utils/facts/base.py
+++ b/test/units/module_utils/facts/base.py
@@ -19,8 +19,10 @@
 from __future__ import (absolute_import, division)
 __metaclass__ = type
 
+import os
+
 from units.compat import unittest
-from units.compat.mock import Mock
+from units.compat.mock import Mock, patch
 
 
 class BaseFactsTest(unittest.TestCase):
@@ -45,14 +47,18 @@ class BaseFactsTest(unittest.TestCase):
         mock_module.get_bin_path = Mock(return_value=None)
         return mock_module
 
-    def test_collect(self):
+    @patch('platform.system', return_value='Linux')
+    @patch('ansible.module_utils.facts.system.service_mgr.get_file_content', return_value='systemd')
+    def test_collect(self, mock_gfc, mock_ps):
         module = self._mock_module()
         fact_collector = self.collector_class()
         facts_dict = fact_collector.collect(module=module, collected_facts=self.collected_facts)
         self.assertIsInstance(facts_dict, dict)
         return facts_dict
 
-    def test_collect_with_namespace(self):
+    @patch('platform.system', return_value='Linux')
+    @patch('ansible.module_utils.facts.system.service_mgr.get_file_content', return_value='systemd')
+    def test_collect_with_namespace(self, mock_gfc, mock_ps):
         module = self._mock_module()
         fact_collector = self.collector_class()
         facts_dict = fact_collector.collect_with_namespace(module=module,

--- a/test/units/module_utils/facts/test_ansible_collector.py
+++ b/test/units/module_utils/facts/test_ansible_collector.py
@@ -207,7 +207,9 @@ class TestCollectedFacts(unittest.TestCase):
     def _mock_module(self, gather_subset=None):
         return mock_module(gather_subset=self.gather_subset)
 
-    def setUp(self):
+    @patch('platform.system', return_value='Linux')
+    @patch('ansible.module_utils.facts.system.service_mgr.get_file_content', return_value='systemd')
+    def setUp(self, mock_gfc, mock_ps):
         mock_module = self._mock_module()
         collectors = self._collectors(mock_module)
 

--- a/test/units/parsing/vault/test_vault.py
+++ b/test/units/parsing/vault/test_vault.py
@@ -247,7 +247,7 @@ class TestFileVaultSecret(unittest.TestCase):
 
     def test_file_not_found(self):
         tmp_file = tempfile.NamedTemporaryFile()
-        filename = tmp_file.name
+        filename = os.path.realpath(tmp_file.name)
         tmp_file.close()
 
         fake_loader = DictDataLoader({filename: 'sdfadf'})
@@ -390,7 +390,7 @@ class TestGetFileVaultSecret(unittest.TestCase):
 
     def test_file_not_found(self):
         tmp_file = tempfile.NamedTemporaryFile()
-        filename = tmp_file.name
+        filename = os.path.realpath(tmp_file.name)
         tmp_file.close()
 
         fake_loader = DictDataLoader({filename: 'sdfadf'})

--- a/test/units/parsing/vault/test_vault_editor.py
+++ b/test/units/parsing/vault/test_vault_editor.py
@@ -492,7 +492,7 @@ class TestVaultEditor(unittest.TestCase):
         self.assertEqual(res, '/dev/null')
 
     def test_real_path_symlink(self):
-        self._test_dir = self._create_test_dir()
+        self._test_dir = os.path.realpath(self._create_test_dir())
         file_path = self._create_file(self._test_dir, 'test_file', content=b'this is a test file')
         file_link_path = os.path.join(self._test_dir, 'a_link_to_test_file')
 

--- a/test/units/playbook/test_helpers.py
+++ b/test/units/playbook/test_helpers.py
@@ -62,7 +62,8 @@ class MixinForMocks(object):
 
         self.mock_block = MagicMock(name='MockBlock')
 
-        self.fake_role_loader = DictDataLoader({"/etc/ansible/roles/bogus_role/tasks/main.yml": """
+        # On macOS /etc is actually /private/etc, tests fail when performing literal /etc checks
+        self.fake_role_loader = DictDataLoader({os.path.join(os.path.realpath("/etc"), "ansible/roles/bogus_role/tasks/main.yml"): """
                                                 - shell: echo 'hello world'
                                                 """})
 

--- a/test/units/utils/test_encrypt.py
+++ b/test/units/utils/test_encrypt.py
@@ -15,6 +15,8 @@
 # You should have received a copy of the GNU General Public License
 # along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
 
+import sys
+
 import pytest
 
 from ansible.errors import AnsibleError, AnsibleFilterError
@@ -46,6 +48,7 @@ def assert_hash(expected, secret, algorithm, **settings):
         assert excinfo.value.args[0] == "passlib must be installed to hash with '%s'" % algorithm
 
 
+@pytest.mark.skipif(sys.platform.startswith('darwin'), reason='macOS requires passlib')
 def test_encrypt_with_rounds_no_passlib():
     with passlib_off():
         assert_hash("$5$12345678$uAZsE3BenI2G.nA8DpTl.9Dc8JiqacI53pEqRr5ppT7",
@@ -67,6 +70,7 @@ def test_encrypt_with_rounds():
                 secret="123", algorithm="sha512_crypt", salt="12345678", rounds=5000)
 
 
+@pytest.mark.skipif(sys.platform.startswith('darwin'), reason='macOS requires passlib')
 def test_encrypt_default_rounds_no_passlib():
     with passlib_off():
         assert_hash("$1$12345678$tRy4cXc3kmcfRZVj4iFXr/",
@@ -89,9 +93,10 @@ def test_encrypt_default_rounds():
     assert_hash("$6$12345678$LcV9LQiaPekQxZ.OfkMADjFdSO2k9zfbDQrHPVcYjSLqSdjLYpsgqviYvTEP/R41yPmhH3CCeEDqVhW1VHr3L.",
                 secret="123", algorithm="sha512_crypt", salt="12345678")
 
-    assert encrypt.CryptHash("md5_crypt").hash("123")
+    assert encrypt.PasslibHash("md5_crypt").hash("123")
 
 
+@pytest.mark.skipif(sys.platform.startswith('darwin'), reason='macOS requires passlib')
 def test_password_hash_filter_no_passlib():
     with passlib_off():
         assert not encrypt.PASSLIB_AVAILABLE
@@ -127,6 +132,7 @@ def test_password_hash_filter_passlib():
     assert get_encrypted_password("123", "pbkdf2_sha256")
 
 
+@pytest.mark.skipif(sys.platform.startswith('darwin'), reason='macOS requires passlib')
 def test_do_encrypt_no_passlib():
     with passlib_off():
         assert not encrypt.PASSLIB_AVAILABLE

